perf(preview): Split preview data to new table

CarlSchwan opened the pull request:
* Resolves: #54513

## Summary

The new oc_previews table is optimized for storing previews and should decrease significantly the space taken by previews in the filecache table.

This attend to reuse the IObjectStore abstraction over S3/Swift/Azure but currently only support one single bucket configuration.

## TODO

- [x] Implement multibucket configuration
- [x] Implement multibucket configuration where previews are stored in different bucket than the original files
- [x] Add migration for local storage
- [x] Add migration for object store storage
- [x] Add migration for multibucket object store storage
- [x] Make migration efficient
- [x] Adapt/Add unit tests
- [x] Re-thing BackgroundCleanupJob and other jobs and commands

## TODO moved to seperate MR

- [ ] Sharding: we want to do the sharding on the file_id table. Likely could be done with a tool like citus or vitess. Need some research
- [ ] Server side encryption: currently we don't use SSE on the previews. If we end up implementing SSE for previews, we should use the same key file as the original files and add a seperate IPreviewStorage implementation using the IEncryptionModule. I already added an encrypted bool to the oc_previews table in preparation as it would be needed anyway for the migration

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] ~~Screenshots before/after for front-end changes~~ Should not contains any visual changes ;)
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] ~~[Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)~~

Commit:
perf(preview): Split preview data to new table

The new oc_previews table is optimized for storing previews and should
decrease significantly the space taken by previews in the filecache
table.

This attend to reuse the IObjectStore abstraction over S3/Swift/Azure
but currently only support one single bucket configuration.

Signed-off-by: Carl Schwan <carl.schwan@nextclound.com>

Altahrim reviewed (commented):
Quick thoughts on the new table

CarlSchwan commented on core/Migrations/Version33000Date20250819110529.php:
```diff
@@ -0,0 +1,51 @@
+			$table->addColumn('etag', Types::STRING, ['notnull' => true, 'length' => 40]);
+			$table->addColumn('mtime', Types::INTEGER, ['notnull' => true]);
+			$table->addColumn('size', Types::INTEGER, ['notnull' => true]);
+			$table->addColumn('version', Types::BIGINT, ['notnull' => false, 'unsigned' => true]);
```
the version is the timestamp as used by the files_versions_s3 app

Altahrim replied:
timestamp should hold in a INTEGER UNSIGNED field

Altahrim commented on core/Migrations/Version33000Date20250819110529.php:
```diff
@@ -0,0 +1,51 @@
+			$table->addColumn('width', Types::INTEGER, ['notnull' => true]);
+			$table->addColumn('height', Types::INTEGER, ['notnull' => true]);
+			$table->addColumn('mimetype', Types::INTEGER, ['notnull' => true]);
+			$table->addColumn('is_max', Types::BOOLEAN, ['notnull' => false]);
+			$table->addColumn('crop', Types::BOOLEAN, ['notnull' => false]);
```
Maybe it works with a default value?

Sadly, it seems to use a `TINYINT(1)` for MySQL. It could also be a solution but will impact SQLite and Postgres which use a BOOLEAN type.

Altahrim replied:
OK, Oracle is the magical word, I give up :D
Nullable boolean seems to be the best compromise

Commit:
perf(preview): Add support for multibucket storage

Signed-off-by: Carl Schwan <carl.schwan@nextclound.com>

Commit:
perf(preview): Migrate previews to the new optimized table

Signed-off-by: Carl Schwan <carl.schwan@nextcloud.com>

CarlSchwan committed:
feat(preview): Support multibucket storage

Signed-off-by: Carl Schwan <carl.schwan@nextcloud.com>

CarlSchwan committed:
perf(preview): Adapt BackgroundCleanupJob to new previews table

Signed-off-by: Carl Schwan <carl.schwan@nextcloud.com>

CarlSchwan commented:
@icewind1991 some questions for you :)

- With #52786 there is now the possibility of using multiple backends per user. The current mapping is done using user configs (oc_preferences) and previews are still stored in the 'root' object store. Should we add a column to the preview table for the backend used (e.g. 'default', 'new', 'old', 'root', ...) to allow an easier migration of the previews in the future? Or do we keeping using only the 'root' backend for previews?

- Multi bucket support. Currently either all the previews are stored on the root bucket or with an opt-in flag can be spread out in 256. I keep the exact same mapping in the new system, does it make sense? Or do we want to use the opportunity of having  a breaking change to change that?

- Sharding: I added a storage_id field to the column with sharding in mind. I didn't configure yet anything else or test it. Should I had it to SHARD_PRESETS in Connection.php?

- Generally how do you feel about the direction this is going?

CarlSchwan committed:
perf(preview): Optimize migration and simplify DB layout

* Simplify migration by not moving the actual files and just updating
  the DB
* Don't store the storageid in the preview table as it is not needed
* Start adding tests

Signed-off-by: Carl Schwan <carl.schwan@nextcloud.com>

CarlSchwan committed:
fix(preview): Fix some tests

Signed-off-by: Carl Schwan <carl.schwan@nextcloud.com>

CarlSchwan committed:
refactor(preview): Cleanup the implementation of the new preview backend

Signed-off-by: Carl Schwan <carl.schwan@nextcloud.com>

artonge reviewed (commented):
~~- Do we have a way to garbage collect orphan previews?~~ Was answered during our call.

provokateurin commented:
Cool, then I probably overlooked that when I checked the migration.

artonge reviewed (commented):
Two questions that I wanted to ask during the review:

1. What happens when a preview that was not migrated yet is requested? Is there a fallback to the legacy system? Or is it generated?
2. Is the preview folder excluded from scan for local storage?

CarlSchwan committed:
feat(preview): Implement scanning for previews

This work similarly to the move preview job to migrate the previews to
the new DB table and also reuse some code.

So when we are finding files in appdata/preview, try adding them to the
oc_previews table and delete them from the oc_filecache table.

Signed-off-by: Carl Schwan <carl.schwan@nextcloud.com>

CarlSchwan committed:
feat(preview): Store original file mimetype in preview table

Allow to quickly query all the files from a specific mimetype like in
the ResetRenderedTexts command.

Signed-off-by: Carl Schwan <carl.schwan@nextcloud.com>

CarlSchwan committed:
refactor: Use Override annotation in new preview code

Signed-off-by: Carl Schwan <carl.schwan@nextcloud.com>

icewind1991 commented on core/BackgroundJobs/MovePreviewJob.php:
```diff
@@ -0,0 +1,209 @@
+					$this->storageFactory->migratePreview($preview, $file);
+					$qb->delete('filecache')
+						->where($qb->expr()->eq('fileid', $qb->createNamedParameter($file->getId())))
+						->runAcrossAllShards() // Unavoidable because we can't extract the storage_id from the preview name
```
same here

CarlSchwan committed:
refactor(preview): Use same mimetype ids as filecache

Signed-off-by: Carl Schwan <carl.schwan@nextcloud.com>

CarlSchwan committed:
fix(preview): Make version column a string

And move it to a different table so that we don't have to pay the
storage cost when not using it (most of the times).

Signed-off-by: Carl Schwan <carl.schwan@nextcloud.com>

CarlSchwan committed:
refactor(preview-object-store): Refactor object store backend

Simplify logic

Signed-off-by: Carl Schwan <carl.schwan@nextcloud.com>

CarlSchwan commented on lib/private/Preview/Storage/ObjectStorePreviewStorage.php:
```diff
@@ -0,0 +1,162 @@
+			$locationId = $this->previewMapper->getLocationId($bucketName, $objectStoreName);
+			$preview->setLocationId($locationId);
+			$preview->setObjectStoreName($objectStoreName);
+			$preview->setBucketName($bucketName);
```
I refactored the whole code, this method is now split in three methods. This particular code is only called when requesting the object store information for a new preview (either migrated from the old database or a newly generated preview)

ChristophWurst commented on core/Migrations/Version33000Date20250819110529.php:
```diff
@@ -0,0 +1,73 @@
+		if (!$schema->hasTable('preview_versions')) {
+			$table = $schema->createTable('preview_versions');
+			$table->addColumn('id', Types::BIGINT, ['autoincrement' => true, 'notnull' => true, 'length' => 20, 'unsigned' => true]);
+			$table->addColumn('file_id', Types::BIGINT, ['notnull' => true, 'length' => 20, 'unsigned' => true]);
```
The mapper confirms that an index on file_id+version would be beneficial :)

keestux commented on core/Migrations/Version33000Date20250819110529.php:
```diff
@@ -0,0 +1,73 @@
+/**
+ *
+ */
+#[CreateTable(table: 'preview', description: 'Holds the preview data')]
```
`preview` or `previews` ?

keestux commented on core/Migrations/Version33000Date20250819110529.php:
```diff
@@ -0,0 +1,73 @@
+ *
+ */
+#[CreateTable(table: 'preview', description: 'Holds the preview data')]
+#[CreateTable(table: 'preview_locations', description: 'Holds the preview location in an object store')]
```
We're missing `preview_versions`
